# Классификациия комментариев для компании Самокат

Компания Самокат обратилась с задачей множественной классификации комментариев к заказам из приложения по заданным темам.
В задании представлены ответы на опрос, состоящий из части с выбором ответа из списка и расширенного комментария с произвольным текстом. Необходимо для каждого ответа из 50 различных меток классов выбрать все затронутые.

**Метрикой** в данной задаче является Accuracy, которая в множественной классификации считается через полное совпадение списка выбранных классов для каждого экземпляра. 

**Описание данных**:  
Исходные данные размещены в файле *train.csv*, тестовые данные в файле *test.csv*.  
Описание искомых категорий указано в файле *trends_description.csv*.  
Также есть пример отправки  результата в файле *sample_submission.csv* и базовая модель классиикации в файле *baseline.ipynb*. 

<a id='К-содержанию'></a>
**Содержание проекта** \
[Загрузка и предобработка данных](#Загрузка-и-предобработка-данных)\
[Подготовка и обучение моделей на векторах TF-IDF](#Подготовка-и-обучение-моделей-на-векторах-TF-IDF)\
[Подготовка и обучение моделей на эмбеддингах Word2Vec](#Подготовка-и-обучение-моделей-на-эмбеддингах-Word2Vec)\
[Подготовка и обучение моделей на эмбеддингах LaBSE](#Подготовка-и-обучение-моделей-на-эмбеддингах-LaBSE)\
[Проверка лучшей модели на валидационной выборке](#Проверка-лучшей-модели-на-валидационной-выборке)\
[Дообучение модели LaBSE для классификации](#Дообучение-модели-LaBSE-для-классификации)\
[Общий вывод](#Общий-вывод)

In [1]:
%%capture
!pip install spacy
!python -m spacy download ru_core_news_sm
!pip install gensim
!pip install sentence_transformers

In [2]:
#импортируем библиотеки перед началом проекта одним блоком
import os
import pandas as pd
import numpy as np
import random
import re
from tqdm.notebook import tqdm
import joblib


# загружаем класс pipeline и классы дя подготовки данных и запуска пайплайна
from sklearn.pipeline import Pipeline
from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from tqdm import notebook

# загружаем нужные модели и метрики
from sklearn.linear_model import LogisticRegression
from sklearn.multioutput import MultiOutputClassifier
from catboost import CatBoostClassifier
from sklearn.model_selection import ParameterGrid
from sklearn.metrics import accuracy_score
from sklearn.base import BaseEstimator, ClassifierMixin

# загружаем функции для подготовки текстовых данных
import nltk
from nltk.tokenize import word_tokenize
import spacy
from gensim.models import Word2Vec
from sklearn.feature_extraction.text import TfidfVectorizer, CountVectorizer 

# загружаем библиотеки для Pytorch и предобученных моделей
import torch
from torch import nn
import torch.optim as optim
import torch.nn.functional as F
import transformers 
from sentence_transformers import SentenceTransformer, util
from transformers import AutoTokenizer, AutoModel, Trainer, TrainingArguments

In [3]:
# зададим стандарт датафрейма перед загрузкой
pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_rows', None)
pd.set_option('display.max_columns', None)

os.environ["TOKENIZERS_PARALLELISM"] = "false"

# устанавливаем фиксированное значение для всех случайных библиотек
RANDOM_STATE = 42

def seed_all(seed_value):
    random.seed(seed_value)
    np.random.seed(seed_value)
    torch.manual_seed(seed_value)
    if torch.cuda.is_available():
        torch.cuda.manual_seed(seed_value)
        torch.cuda.manual_seed_all(seed_value)
        torch.backends.cudnn.benchmark = True
        torch.backends.cudnn.deterministic = False
seed_all(RANDOM_STATE)

# функции для лемматизаци и исключения ненужных символов в комментариях
nlp = spacy.load("ru_core_news_sm")

def cleaner(text: str) -> str:
    
    text = re.sub(r'\_(,*?)\_', '', text)
    text = re.sub(r'[\n\t]', '', text)
    text = re.sub(r'<[^>]*>', '', text)
    text = re.sub(r'http\S+|www\S+|https\S+', '', text)
    text = " ".join(text.split())
    return text

def lemmatize(text):
    doc = nlp(text)
    lemm_list = [token.lemma_ for token in doc if not token.is_punct] # сразу убираем пунктуацию 
    lemm_text = " ".join(lemm_list)
    return lemm_text

def process_text(text):
    text = cleaner(text)
    text = lemmatize(text)
    return text

<a id='Загрузка-и-предобработка-данных'></a> 
## Загрузка и предобработка данных
[К содержанию](#К-содержанию)

In [4]:
# импортируем тренировочные данные
pth1 = 'Data/train.csv'

if os.path.exists(pth1):
    data = pd.read_csv(pth1)
else:
    print('Something is wrong')

data.info()
display(data.head())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8708 entries, 0 to 8707
Data columns (total 54 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   index           8708 non-null   int64  
 1   assessment      8478 non-null   float64
 2   tags            8697 non-null   object 
 3   text            8708 non-null   object 
 4   trend_id_res0   8708 non-null   int64  
 5   trend_id_res1   8708 non-null   int64  
 6   trend_id_res2   8708 non-null   int64  
 7   trend_id_res3   8708 non-null   int64  
 8   trend_id_res4   8708 non-null   int64  
 9   trend_id_res5   8708 non-null   int64  
 10  trend_id_res6   8708 non-null   int64  
 11  trend_id_res7   8708 non-null   int64  
 12  trend_id_res8   8708 non-null   int64  
 13  trend_id_res9   8708 non-null   int64  
 14  trend_id_res10  8708 non-null   int64  
 15  trend_id_res11  8708 non-null   int64  
 16  trend_id_res12  8708 non-null   int64  
 17  trend_id_res13  8708 non-null   i

index  assessment  \
0   5652         6.0   
1  18092         4.0   
2  13845         6.0   
3  25060         6.0   
4  15237         5.0   

                                                          tags  \
0                             {ASSORTMENT,PROMOTIONS,DELIVERY}   
1                 {ASSORTMENT,PRICE,PRODUCTS_QUALITY,DELIVERY}   
2               {DELIVERY,PROMOTIONS,PRICE,ASSORTMENT,SUPPORT}   
3                                {PRICE,PROMOTIONS,ASSORTMENT}   
4  {ASSORTMENT,PRODUCTS_QUALITY,PROMOTIONS,CATALOG_NAVIGATION}   

                                                                                 text  \
0  Маленький выбор товаров, хотелось бы ассортимент больше, а так вроде бы все хорошо   
1                                                                              Быстро   
2                                                    Доставка постоянно задерживается   
3                                                  Наценка и ассортимент расстраивают   
4                                                                   Доставка просто 👍   

   trend_id_res0  trend_id_res1  trend_id_res2  trend_id_res3  trend_id_res4  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              1              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   trend_id_res5  trend_id_res6  trend_id_res7  trend_id_res8  trend_id_res9  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   trend_id_res10  trend_id_res11  trend_id_res12  trend_id_res13  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               1               0   
4               0               0               0               0   

   trend_id_res14  trend_id_res15  trend_id_res16  trend_id_res17  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               0               0               0               0   

   trend_id_res18  trend_id_res19  trend_id_res20  trend_id_res21  \
0               0               0               0               0   
1               1               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               0               0               0               0   

   trend_id_res22  trend_id_res23  trend_id_res24  trend_id_res25  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               0               0               0               0   

   trend_id_res26  trend_id_res27  trend_id_res28  trend_id_res29  \
0               0               0               1               0   
1               0               0               0               0   
2               0               0               0               0   
3               0           

In [5]:
# проверим целевые столбцы на нулевые значения - нет смысла тренировать модель на пустых
target_columns = [f'trend_id_res{i}' for i in range(50)]
target_all_zeros = (data[target_columns] == 0).all(axis=1)
num_target_all_zeros = target_all_zeros.sum()

print(f"Количество строк с нулями в целевых столбцах: {num_target_all_zeros}")

Количество строк с нулями в целевых столбцах: 4085


С точки зрения обучения моделей данные строки не имеют смысла, однако, наборы текстов будут полезны для обучения эмбеддингов, поэтому оставим.

In [6]:
#отфильтруем и уберем строки, где все значения целевых столбцов равны 0
#data = data[~(data[target_columns] == 0).all(axis=1)]
#data = data.reset_index(drop=True)
#print("Строки с нулями удалены.")

In [7]:
# уберем столбец индекса отзыва
data = data.drop('index',axis=1)

In [8]:
# проверим дубликаты
print('Количество дубликатов: ', data.duplicated().sum().sum())

# проверим пропуски
print('Количество пропусков: ', data.isna().sum().sum())

# уберем дубликаты и пропуски
data = data.drop_duplicates()
data = data.dropna()

Количество дубликатов:  246
Количество пропусков:  241


In [9]:
# преобразуем значения категорий в столбце tags
data['tags'] = data['tags'].str.strip('{}').str.lower().str.split(',')
data.head()

assessment                                                            tags  \
0         6.0                              [assortment, promotions, delivery]   
1         4.0                 [assortment, price, products_quality, delivery]   
2         6.0              [delivery, promotions, price, assortment, support]   
3         6.0                                 [price, promotions, assortment]   
4         5.0  [assortment, products_quality, promotions, catalog_navigation]   

                                                                                 text  \
0  Маленький выбор товаров, хотелось бы ассортимент больше, а так вроде бы все хорошо   
1                                                                              Быстро   
2                                                    Доставка постоянно задерживается   
3                                                  Наценка и ассортимент расстраивают   
4                                                                   Доставка просто 👍   

   trend_id_res0  trend_id_res1  trend_id_res2  trend_id_res3  trend_id_res4  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              1              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   trend_id_res5  trend_id_res6  trend_id_res7  trend_id_res8  trend_id_res9  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   trend_id_res10  trend_id_res11  trend_id_res12  trend_id_res13  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               1               0   
4               0               0               0               0   

   trend_id_res14  trend_id_res15  trend_id_res16  trend_id_res17  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               0               0               0               0   

   trend_id_res18  trend_id_res19  trend_id_res20  trend_id_res21  \
0               0               0               0               0   
1               1               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               0               0               0               0   

   trend_id_res22  trend_id_res23  trend_id_res24  trend_id_res25  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               0               0               0               0   

   trend_id_res26  trend_id_res27  trend_id_res28  trend_id_res29  \
0               0               0               1               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               1               0   
4             

In [10]:
# собираем все категории в один список
all_categories = data['tags'].explode()

# смотрим уникальные значения категорий
all_categories.unique()

array(['assortment', 'promotions', 'delivery', 'price',
       'products_quality', 'support', 'catalog_navigation', 'payment'],
      dtype=object)

In [11]:
# категорий всего 8, можем закодировать их бинарно, создав 8 столбцов

for category in all_categories:
    data[f'tags_{category}'] = data['tags'].apply(lambda x: 1 if category in x else 0)

data = data.drop('tags', axis=1)

data.head()

assessment  \
0         6.0   
1         4.0   
2         6.0   
3         6.0   
4         5.0   

                                                                                 text  \
0  Маленький выбор товаров, хотелось бы ассортимент больше, а так вроде бы все хорошо   
1                                                                              Быстро   
2                                                    Доставка постоянно задерживается   
3                                                  Наценка и ассортимент расстраивают   
4                                                                   Доставка просто 👍   

   trend_id_res0  trend_id_res1  trend_id_res2  trend_id_res3  trend_id_res4  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              1              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   trend_id_res5  trend_id_res6  trend_id_res7  trend_id_res8  trend_id_res9  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   trend_id_res10  trend_id_res11  trend_id_res12  trend_id_res13  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               1               0   
4               0               0               0               0   

   trend_id_res14  trend_id_res15  trend_id_res16  trend_id_res17  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               0               0               0               0   

   trend_id_res18  trend_id_res19  trend_id_res20  trend_id_res21  \
0               0               0               0               0   
1               1               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               0               0               0               0   

   trend_id_res22  trend_id_res23  trend_id_res24  trend_id_res25  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               0               0               0               0   

   trend_id_res26  trend_id_res27  trend_id_res28  trend_id_res29  \
0               0               0               1               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               1               0   
4               0               0               0               0   

   trend_id_res30  trend_id_res31  trend_id_res32  trend_id_res33  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               0   

In [12]:
print(f'Пример исходного текста тренировочной выборки: \n\n {data.iloc[0,1]} \n')

Пример исходного текста тренировочной выборки: 

 Маленький выбор товаров, хотелось бы ассортимент больше, а так вроде бы все хорошо 



In [13]:
# запускаем tqdm для отображения прогресса
tqdm.pandas()

data_lem = data.copy()

# выполняем лемматизацию с отображением прогресса
data_lem['text'] = data_lem['text'].progress_apply(lambda x: process_text(x))

# удалим пустые после лемматизации строки и соответствующие им значения целевого признака
data_lem = data_lem.dropna().reset_index(drop=True)

  0%|          | 0/8222 [00:00<?, ?it/s]

In [14]:
print(f'Пример преобразованного текста тренировочной выборки: \n\n {data_lem.iloc[0,1]} \n')

Пример преобразованного текста тренировочной выборки: 

 маленький выбор товар хотеться бы ассортимент большой а так вроде бы все хороший 



In [15]:
# добавим признак длины текста
data_lem['text_len'] = data_lem['text'].apply(lambda x: len(x.split()))

# добавим количества тегов
data_lem['tags_count'] = data_lem.filter(like='tags_').sum(axis=1)

data_lem.head()

assessment  \
0         6.0   
1         4.0   
2         6.0   
3         6.0   
4         5.0   

                                                                               text  \
0  маленький выбор товар хотеться бы ассортимент большой а так вроде бы все хороший   
1                                                                            Быстро   
2                                                  доставка постоянно задерживаться   
3                                                наценка и ассортимент расстраивать   
4                                                                 доставка просто 👍   

   trend_id_res0  trend_id_res1  trend_id_res2  trend_id_res3  trend_id_res4  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              1              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   trend_id_res5  trend_id_res6  trend_id_res7  trend_id_res8  trend_id_res9  \
0              0              0              0              0              0   
1              0              0              0              0              0   
2              0              0              0              0              0   
3              0              0              0              0              0   
4              0              0              0              0              0   

   trend_id_res10  trend_id_res11  trend_id_res12  trend_id_res13  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               1               0   
4               0               0               0               0   

   trend_id_res14  trend_id_res15  trend_id_res16  trend_id_res17  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               0               0               0               0   

   trend_id_res18  trend_id_res19  trend_id_res20  trend_id_res21  \
0               0               0               0               0   
1               1               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               0               0               0               0   

   trend_id_res22  trend_id_res23  trend_id_res24  trend_id_res25  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               0               0   
4               0               0               0               0   

   trend_id_res26  trend_id_res27  trend_id_res28  trend_id_res29  \
0               0               0               1               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               1               0   
4               0               0               0               0   

   trend_id_res30  trend_id_res31  trend_id_res32  trend_id_res33  \
0               0               0               0               0   
1               0               0               0               0   
2               0               0               0               0   
3               0               0               0               

In [16]:
# сделаем индикатор "пустых" строк (где все значения равны 0)
is_empty = (data_lem[target_columns].sum(axis=1) == 0).astype(int)

# разделим выборку на тренировочную и тестовую с учетом стратификации по "пустым" признакам
X_train, X_test, y_train, y_test = train_test_split(
    data_lem.drop(target_columns,axis=1), data_lem[target_columns], test_size=0.15, random_state=RANDOM_STATE, stratify=is_empty)

X_train = X_train.reset_index(drop=True)
X_test = X_test.reset_index(drop=True)
y_train = y_train.reset_index(drop=True)
y_test = y_test.reset_index(drop=True)

# проверим размерности после разбивки
print('Размерности датасетов после разбивки: \n',
    X_train.shape,
    X_test.shape,
    y_train.shape,
    y_test.shape)


Размерности датасетов после разбивки: 
 (6988, 12) (1234, 12) (6988, 50) (1234, 50)


In [17]:
# и преобразуем столбцы целевой переменной в массив numpy - для некоторых моделей критично
y_train = y_train.to_numpy()
y_test = y_test.to_numpy()
display(y_train[:2],
        y_test[:2])

array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0],
       [0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0]])

array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
        0, 0, 0, 0, 0, 0]])

In [18]:
flag = 0

<a id='#Подготовка-и-обучение-моделей-на-векторах-TF-IDF'></a> 
## Подготовка и обучение моделей на векторах TF-IDF
[К содержанию](#К-содержанию)


Сначала попробуем самый простой вариант (на базе baseline модели) с применением векторов TF-IDF:

In [19]:
# закодируем комментарии с помощью tf-idf
X_train_tf_idf = X_train.copy()

# преобразуем тексты в tf-idf векторы
tf_idf_vectorizer = TfidfVectorizer() #analyzer="char_wb",ngram_range=(2,3))#,max_features=5000)
tf_idf_vectors = tf_idf_vectorizer.fit_transform(X_train_tf_idf['text'])

# преобразуем разреженную матрицу в DataFrame
tf_idf_df = pd.DataFrame(
    tf_idf_vectors.toarray(),
    columns=tf_idf_vectorizer.get_feature_names_out(),  # берем имена признаков
    index=X_train_tf_idf.index  # индексы для соответствия
)

# объединяем исходный датафрейм с признаками tf-idf
X_train_tf_idf = pd.concat([X_train_tf_idf, tf_idf_df], axis=1)

# удаляем ненужный столбец с текстами
X_train_tf_idf = X_train_tf_idf.drop(columns=['text'])

# выведем пару строк для проверки
X_train_tf_idf.head(2)

assessment  tags_assortment  tags_promotions  tags_delivery  tags_price  \
0         2.0                0                1              0           0   
1         6.0                1                1              1           0   

   tags_products_quality  tags_support  tags_catalog_navigation  tags_payment  \
0                      0             1                        0             0   
1                      0             1                        0             0   

   text_len  tags_count  always  cola  darkstore  \
0        11           2     0.0   0.0        0.0   
1        17           4     0.0   0.0        0.0   

   devochkavensdeigrustniemojitimnenravishsyatemchtoninakogonepohozha  done  \
0                                                                 0.0   0.0   
1                                                                 0.0   0.0   

   elit  fps  horosho  horror  huggies  jambo  klasss  knows  kruto  luchse  \
0   0.0  0.0      0.0     0.0      0.0    0.0     0.0    0.0    0.0     0.0   
1   0.0  0.0      0.0     0.0      0.0    0.0     0.0    0.0    0.0     0.0   

   mozhno  nestea   no       num  nuts  phone  pixel  pro  sale  soft  super  \
0     0.0     0.0  0.0  0.000000   0.0    0.0    0.0  0.0   0.0   0.0    0.0   
1     0.0     0.0  0.0  0.140885   0.0    0.0    0.0  0.0   0.0   0.0    0.0   

   top   up  vhs  well  who   wk  абонемент  абсолютно  абузить  авангард  \
0  0.0  0.0  0.0   0.0  0.0  0.0        0.0        0.0      0.0       0.0   
1  0.0  0.0  0.0   0.0  0.0  0.0        0.0        0.0      0.0       0.0   

   авка  авокадо  авто  автомат  автоматизация  автоматически   ад  адекват  \
0   0.0      0.0   0.0      0.0            0.0            0.0  0.0      0.0   
1   0.0      0.0   0.0      0.0            0.0            0.0  0.0      0.0   

   адекватно  адекватный  аджарски  администратор  адреналин  адрес  адресу  \
0        0.0         0.0       0.0            0.0        0.0    0.0     0.0   
1        0.0         0.0       0.0            0.0        0.0    0.0     0.0   

   адский   аж  аза  азбука  азбуке  азиатский  айс  айфон  айчары  академа  \
0     0.0  0.0  0.0     0.0     0.0        0.0  0.0    0.0     0.0      0.0   
1     0.0  0.0  0.0     0.0     0.0        0.0  0.0    0.0     0.0      0.0   

   аккаунт  аккуратно  аксессуар  активно  активность  актуальный  акцент  \
0      0.0        0.0        0.0      0.0         0.0         0.0     0.0   
1      0.0        0.0        0.0      0.0         0.0         0.0     0.0   

   акционные  акционных  акция  алко  алкоголь  алкогольный  альтернатива  \
0        0.0        0.0    0.0   0.0       0.0          0.0           0.0   
1        0.0        0.0    0.0   0.0       0.0          0.0           0.0   

   альтернативный  альтернативу  аля   ам  анализировать  аналог  аналогичный  \
0             0.0           0.0  0.0  0.0            0.0     0.0          0.0   
1             0.0           0.0  0.0  0.0            0.0     0.0          0.0   

   ананас  ангел  андроид  анигири  анус  апельсин  апельсинов  аподуктов  \
0     0.0    0.0      0.0      0.0   0.0       0.0         0.0        0.0   
1     0.0    0.0      0.0      0.0   0.0       0.0         0.0        0.0   

   аппетит  аптека  аргументировать  ареал  аромат  асортимент  асортиментов  \
0      0.0     0.0              0.0    0.0     0.0         0.0           0.0   
1      0.0     0.0              0.0    0.0     0.0         0.0           0.0   

   ассорти  ассортимент  ассортимента  ассортименту  ассорьимент  асфальт  \
0      0.0          0.0           0.0           0.0          0.0      0.0   
1      0.0          0.0           0.0           0.0          0.0      0.0   

   атас  атмосфера  аудитория  ахуели  ахуенно  ашан  аыышенные  бабка  \
0   0.0        0.0        0.0     0.0      0.0   0.0        0.0    0.0   
1   0.0        0.0        0.0     0.0      0.0   0.0        0.0    0.0   

   бабуля  бабушка  бабушкам  баг  багет  багуется  база  базовый  б

In [20]:
# посмотрим на размерность матрицы признаков
X_train_tf_idf.shape

(6988, 5509)

In [21]:
# создаём итоговый пайплайн с выбором модели и гиперпараметров для обучения на tf-idf
pipe_final_tfidf = Pipeline([
    ('models', MultiOutputClassifier(LogisticRegression(random_state=RANDOM_STATE)))
])

# Задаем наборы гиперпараметров для перебора
param_grid_tfidf = [
    # Словарь для модели LogisticRegression
    {
        'models': [MultiOutputClassifier(LogisticRegression(random_state=RANDOM_STATE))],
        'models__estimator__max_iter': [1000],
        'models__estimator__C': [10],
        'models__estimator__penalty': ['l2'],
        'models__estimator__class_weight': [None]
    },
    
    # Словарь для модели CatBoostClassifier
    {
        'models': [MultiOutputClassifier(CatBoostClassifier(silent=True, random_state=RANDOM_STATE, auto_class_weights='Balanced'))],
        'models__estimator__depth': [7],
        'models__estimator__iterations': [150],
        'models__estimator__learning_rate': [0.1],
        'models__estimator__l2_leaf_reg': [3]
    },
]

In [22]:
# полный перебор гиперпараметров с помощью GridSearchCV с применением tqdm

param_grid_tfidf = ParameterGrid(param_grid_tfidf)  # параметры для перебора внутри tqdm

if flag != 1:
    # настройка tqdm
    total_combinations = len(param_grid_tfidf)
    with tqdm(total=total_combinations) as pbar:
        best_score = -float('inf')
        best_params = None
        best_model = None


        for params in param_grid_tfidf:
            grid_search_tfidf = GridSearchCV(
                pipe_final_tfidf,
                {k: [v] for k, v in params.items()},  # подставляем конкретный набор параметров
                cv=5,
                scoring='accuracy',
                n_jobs=-1,
            )
            model_tfidf = grid_search_tfidf.fit(X_train_tf_idf, y_train)

            print(f'Метрика Accuracy: {model_tfidf.best_score_}')
            print(f'Параметры модели: {model_tfidf.best_params_}\n')

            # Обновление лучшей метрики
            if model_tfidf.best_score_ > best_score:
                best_score = model_tfidf.best_score_
                best_params = params
                best_model = model_tfidf

            pbar.update(1)  # обновляем прогресс бар tqdm

    # выгружаем лучшую метрику accuracy и параметры модели
    print(f'Лучшая метрика Accuracy: {best_score}')
    print(f'Параметры лучшей модели: {best_params}')


  0%|          | 0/2 [00:00<?, ?it/s]

Метрика Accuracy: 0.5113082089865572
Параметры модели: {'models': MultiOutputClassifier(estimator=LogisticRegression(random_state=42)), 'models__estimator__C': 10, 'models__estimator__class_weight': None, 'models__estimator__max_iter': 1000, 'models__estimator__penalty': 'l2'}

Метрика Accuracy: 0.5216117103582888
Параметры модели: {'models': MultiOutputClassifier(estimator=<catboost.core.CatBoostClassifier object at 0x336469c50>), 'models__estimator__depth': 7, 'models__estimator__iterations': 150, 'models__estimator__l2_leaf_reg': 3, 'models__estimator__learning_rate': 0.1}

Лучшая метрика Accuracy: 0.5216117103582888
Параметры лучшей модели: {'models': MultiOutputClassifier(estimator=<catboost.core.CatBoostClassifier object at 0x336469c50>), 'models__estimator__depth': 7, 'models__estimator__iterations': 150, 'models__estimator__l2_leaf_reg': 3, 'models__estimator__learning_rate': 0.1}


<a id='Подготовка-и-обучение-моделей-на-эмбеддингах-Word2Vec'></a> 
## Подготовка и обучение моделей на эмбеддингах Word2Vec
[К содержанию](#К-содержанию)

Далее попробуем усложнить процедуру подготовки текстовых данных - обучить Word2Vec на корпусе текстов, чтобы использовать эмбеддинги для сравнения текстов:

In [23]:
X_train_wv = X_train.copy()

# делаем токенизацию
X_train_wv['tokens'] = X_train_wv['text'].apply(lambda x: word_tokenize(x.lower()))
# тренируем Word2Vec только на тренировочной выборке
model_wv = Word2Vec(sentences=X_train_wv['tokens'], vector_size=150, window=5, min_count=1, workers=1, seed=RANDOM_STATE) 
# преобразуем текст в векторы
X_train_wv['text_wv'] = X_train_wv['tokens'].apply(lambda x: [model_wv.wv[word] for word in x if word in model_wv.wv])
# усредняем векторы и удаляем ненужные столбцы
X_train_wv['text_wv_avg'] = X_train_wv['text_wv'].apply(lambda x: np.mean(x, axis=0) if len(x) > 0 else np.zeros(model_wv.vector_size))
X_train_wv = X_train_wv.drop(columns=['text', 'tokens', 'text_wv'])


In [24]:
# развернем векторы в отдельные столбцы
vector_text = pd.DataFrame(X_train_wv['text_wv_avg'].tolist(), columns=[f'vector_{i}' for i in range(150)])
X_train_wv = pd.concat([X_train_wv.drop(columns=['text_wv_avg']), vector_text], axis=1)

print(X_train_wv.shape)

print(X_train_wv.index)
print(vector_text.index)

display(X_train_wv.shape)
display(X_train_wv.head())


(6988, 161)
RangeIndex(start=0, stop=6988, step=1)
RangeIndex(start=0, stop=6988, step=1)


(6988, 161)

assessment  tags_assortment  tags_promotions  tags_delivery  tags_price  \
0         2.0                0                1              0           0   
1         6.0                1                1              1           0   
2         5.0                1                1              0           0   
3         5.0                1                1              1           1   
4         5.0                1                1              1           0   

   tags_products_quality  tags_support  tags_catalog_navigation  tags_payment  \
0                      0             1                        0             0   
1                      0             1                        0             0   
2                      0             0                        0             0   
3                      1             1                        1             1   
4                      1             0                        0             1   

   text_len  tags_count  vector_0  vector_1  vector_2  vector_3  vector_4  \
0        11           2  0.269420 -0.698599  0.144264 -0.434142  0.005050   
1        17           4  0.243407 -0.713528  0.147451 -0.491553 -0.003491   
2         8           2  0.329595 -0.658940  0.145794 -0.276723  0.003178   
3         6           8  0.238459 -0.500242  0.115526 -0.213891  0.011562   
4         3           5  0.339541 -0.587550  0.132916 -0.194947 -0.011441   

   vector_5  vector_6  vector_7  vector_8  vector_9  vector_10  vector_11  \
0 -0.049000  0.123394  0.356737 -0.109586  0.489625  -0.089212   0.173628   
1 -0.096329  0.119309  0.379456 -0.161355  0.510766  -0.089673   0.221333   
2 -0.093591  0.217396  0.230454 -0.161267  0.407989  -0.143719   0.002921   
3 -0.073511  0.162264  0.178781 -0.120003  0.315455  -0.099921   0.004631   
4 -0.133384  0.252834  0.165655 -0.206497  0.343870  -0.172981  -0.088272   

   vector_12  vector_13  vector_14  vector_15  vector_16  vector_17  \
0   0.214612   0.107784  -0.254091  -0.110134   0.286879  -0.568271   
1   0.274994   0.140361  -0.235619  -0.109876   0.305752  -0.575177   
2   0.155540   0.141680  -0.235428  -0.163041   0.183664  -0.651321   
3   0.120160   0.107761  -0.173283  -0.129829   0.137216  -0.496854   
4   0.143206   0.159879  -0.233068  -0.169136   0.112193  -0.669169   

   vector_18  vector_19  vector_20  vector_21  vector_22  vector_23  \
0  -0.269523  -0.454514   0.114671   0.589652  -0.019308   0.541714   
1  -0.284274  -0.462524   0.117464   0.630233  -0.011851   0.569802   
2  -0.299864  -0.366229   0.083798   0.638771  -0.084105   0.669870   
3  -0.232370  -0.283020   0.070731   0.480177  -0.071558   0.501798   
4  -0.312945  -0.318452   0.043490   0.639565  -0.134502   0.718263   

   vector_24  vector_25  vector_26  vector_27  vector_28  vector_29  \
0   0.467113  -0.017316  -0.370478  -0.252914   0.227380   0.366879   
1   0.512866  -0.005715  -0.340630  -0.255758   0.279953   0.358009   
2   0.491149   0.140131  -0.512208  -0.274256   0.041420   0.486109   
3   0.374416   0.103204  -0.394354  -0.202467   0.026818   0.362503   
4   0.498974   0.228532  -0.572308  -0.280666  -0.044910   0.518362   

   vector_30  vector_31  vector_32  vector_33  vector_34  vector_35  \
0   0.234410  -0.707886  -0.383046   0.187913  -0.136802   0.155774   
1   0.268182  -0.764257  -0.404378   0.200108  -0.112677   0.138314   
2   0.231504  -0.562749  -0.458115   0.099520  -0.233891   0.065484   
3   0.160255  -0.424572  -0.345057   0.079232  -0.179531   0.048874   
4   0.252193  -0.469120  -0.494436   0.049007  -0.245934   0.005541   

   vector_36  vector_37  vector_38  vector_39  vector_40  vector_41  \
0  -0.023496  -0.720888  -0.133556   0.205031  -0.189871   0.142447   
1  -0.022025  -0.760402  -0.124583   0.184441  -0.217967   0.158817   
2  -0.021654  -0.663934  -0.040973   0.360134  -0.204100   0.069699   
3  -0.021255  -0.491994  -0.044837   0.278801  -0.152993   0.056493   
4  -0.008812  -0.619369   0.008214   0.423157  -0.220149  

In [25]:
# создаём итоговый пайплайн с выбором модели и гиперпараметров для обучения на эмбеддингах word2vec 
pipe_final_wv = Pipeline([
    ('models', MultiOutputClassifier(LogisticRegression(random_state=RANDOM_STATE)))
])

# задаем наборы гиперпараметров для перебора - используем аналогичные параметры, что и для tf-idf 
# оставляем лучшие наборы гиперпараметров, которые удалось получить на кросс-валидации - полный перебор опускаем

param_grid_wv = [
    # Словарь для модели LogisticRegression
    {
        'models': [MultiOutputClassifier(LogisticRegression(random_state=RANDOM_STATE))],
        'models__estimator__max_iter': [1000],
        'models__estimator__C': [10],
        'models__estimator__penalty': ['l2'],
        'models__estimator__class_weight': [None]
    },

    # Словарь для модели CatBoostClassifier
    {
        'models': [MultiOutputClassifier(CatBoostClassifier(silent=True, random_state=RANDOM_STATE, auto_class_weights='Balanced'))],
        'models__estimator__depth': [7],
        'models__estimator__iterations': [150],
        'models__estimator__learning_rate': [0.1],
        'models__estimator__l2_leaf_reg': [3],
    }
]

In [26]:
# полный перебор гиперпараметров с помощью GridSearchCV с применением tqdm
if flag != 1:
    param_grid_wv = ParameterGrid(param_grid_wv)  # параметры для перебора внутри tqdm

    # настройка tqdm
    total_combinations = len(param_grid_wv)
    with tqdm(total=total_combinations) as pbar:
        best_score = -float('inf')
        best_params = None
        best_model = None


        for params in param_grid_wv:
            grid_search_wv = GridSearchCV(
                pipe_final_wv,
                {k: [v] for k, v in params.items()},  # подставляем конкретный набор параметров
                cv=5,
                scoring='accuracy',
                n_jobs=-1,
            )
            model_wv = grid_search_wv.fit(X_train_wv, y_train)

            print(f'Метрика Accuracy: {model_wv.best_score_}')
            print(f'Параметры модели: {model_wv.best_params_}\n')

            # Обновление лучшей метрики
            if model_wv.best_score_ > best_score:
                best_score = model_wv.best_score_
                best_params = params
                best_model = model_wv

            pbar.update(1)  # обновляем прогресс бар tqdm

    # выгружаем лучшую метрику Accuracy и параметры модели
    print(f'Лучшая метрика Accuracy: {best_score}')
    print(f'Параметры лучшей модели: {best_params}')


  0%|          | 0/2 [00:00<?, ?it/s]

Метрика Accuracy: 0.48669251400149305
Параметры модели: {'models': MultiOutputClassifier(estimator=LogisticRegression(random_state=42)), 'models__estimator__C': 10, 'models__estimator__class_weight': None, 'models__estimator__max_iter': 1000, 'models__estimator__penalty': 'l2'}

Метрика Accuracy: 0.5635399993650813
Параметры модели: {'models': MultiOutputClassifier(estimator=<catboost.core.CatBoostClassifier object at 0x32e8e3650>), 'models__estimator__depth': 7, 'models__estimator__iterations': 150, 'models__estimator__l2_leaf_reg': 3, 'models__estimator__learning_rate': 0.1}

Лучшая метрика Accuracy: 0.5635399993650813
Параметры лучшей модели: {'models': MultiOutputClassifier(estimator=<catboost.core.CatBoostClassifier object at 0x32e8e3650>), 'models__estimator__depth': 7, 'models__estimator__iterations': 150, 'models__estimator__l2_leaf_reg': 3, 'models__estimator__learning_rate': 0.1}


<a id='Подготовка-и-обучение-моделей-на-эмбеддингах-LaBSE'></a> 
## Подготовка и обучение моделей на эмбеддингах LaBSE
[К содержанию](#К-содержанию)

Далее подключим предобученную модель LaBSE-ru-turbo (на базе BERT) для создания эмбеддингов предложений.

In [27]:
# загружаем модель
model_LaBSE = SentenceTransformer('sergeyzh/LaBSE-ru-turbo')

# копируем тренировочный датасет для возможной предобработки
X_train_bert = X_train.copy()

In [28]:
tqdm.pandas()

# преобразуем текст в векторы
X_train_bert['text_vect'] = X_train_bert['text'].progress_apply(lambda x: model_LaBSE.encode(x))

# удаляем ненужные столбцы
X_train_bert = X_train_bert.drop(columns=['text'])

X_train_bert.head(2)

  0%|          | 0/6988 [00:00<?, ?it/s]

assessment  tags_assortment  tags_promotions  tags_delivery  tags_price  \
0         2.0                0                1              0           0   
1         6.0                1                1              1           0   

   tags_products_quality  tags_support  tags_catalog_navigation  tags_payment  \
0                      0             1                        0             0   
1                      0             1                        0             0   

   text_len  tags_count  \
0        11           2   
1        17           4   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                               text_vect  
0   [0.058266375, 0.0022367619, -0.02366651, 0.022975221, -0.033910956, 0.004358709, 0.0209202, -0.015890272, 0.022248752, 0.014778344, 0.015026953, -0.035889693, -0.0028999206, 0.013294978, 0.045170024, 0.007889779, 0.0004207931, 0.025048347, 0.014167981, -0.042518064, -0.03961556, -0.005228488, -0.0066043083, 0.036100764, 0.04221276, 0.025370171, -0.003304544, -0.0210686, -0.026388539, 0.03415403, -0.017136373, 0.006048383, 0.020480432, -0.019974388, -0.0010280486, -0.0104539525, 0.0035239344, 0.037149735, 0.04785923, -0.06885591, -0.01710669, -0.001799424, -0.0017648645, -0.01939534, -0.003730511, 0.016777463, -0.008671057, 0.006506621, -0.017093657, 0.017092971, -0.02702758, -0.00933273, 0.03854842, 0.006555776, 0.027438173, -0.0074756476, -0.02449172, 0.020876858, -0.004892325, -0.006262041, 0.0007645724, -0.020052182, 0.0031566955, 0.001788244, -0.0027312865, -0.03752136, -0.023035696, 0.020432653, 0.05881169, -0.022733416, -0.007107246, -0.029933728, -0.007101953, 0.001970953, -0.013242534, -0.02896522, 0.008165765, 0.023935718, -0.017479222, 0.034317687, -0.01183603, 0.032513697, 0.017316436, 0.028634626, -0.00062191376, 0.013601478, 0.013038712, -0.014413536, 0.037066307, 0.051547296, -0.0071508978, 0.025472954, -0.0113213565, -0.011264834, 0.049866345, 0.005914229, 0.043773346, -0.05123308, 0.0047616544, 0.02217594, ...]  
1  [-0.010358731, -0.015012443, -0.029044122, 0.017889928, 0.007456112, 0.04176937, 0.03704001, 0.0022503205, -0.008878538, -0.009717544, 0.04383548, -0.0016043839, -0.007494684, 0.04988332, 0.020680266, 0.0018086971, 0.02961041, -0.05341848, 0.0054917857, 0.0365915, -0.04096404, -0.0014320223, -0.02387368, 0.004409981, 0.0070271245, 0.020686949, 0.007594833, 0.010309676, -0.017559292, 0.033824563, 0.013475971, 0.0397598, 0.042391844, -0.023852829, -0.016018573, 0.018561225, -0.0489381, 0.002632114, 0.01272903, -0.046140913, -0.019763567, 0.0055882838, -0.022248046, -0.0039494005, -0.07858931, -0.007885234, -1.0952115e-05, -0.007000204, 0.03041942, 0.0030753936, 0.033023827, 0.0023799024, 0.0031912313, -0.042145308, 0.0674

In [29]:
# определяем количество столбцов в векторе
vector_length = len(X_train_bert['text_vect'].iloc[0])

# генерируем названия столбцов
column_names = [f"text_vect_{i}" for i in range(vector_length)]

# создаем DataFrame с заданными названиями столбцов
text_vect_df = pd.DataFrame(X_train_bert['text_vect'].tolist(), 
                            index=X_train_bert.index, 
                            columns=column_names)

# объединяем с исходным DataFrame
X_train_bert = pd.concat([X_train_bert, text_vect_df], axis=1)

# удаляем колонку с исходным вектором
X_train_bert = X_train_bert.drop(columns=['text_vect'])

In [30]:
display(X_train_bert.head(2))
display(X_train_bert.shape)
y_train.shape

assessment  tags_assortment  tags_promotions  tags_delivery  tags_price  \
0         2.0                0                1              0           0   
1         6.0                1                1              1           0   

   tags_products_quality  tags_support  tags_catalog_navigation  tags_payment  \
0                      0             1                        0             0   
1                      0             1                        0             0   

   text_len  tags_count  text_vect_0  text_vect_1  text_vect_2  text_vect_3  \
0        11           2     0.058266     0.002237    -0.023667     0.022975   
1        17           4    -0.010359    -0.015012    -0.029044     0.017890   

   text_vect_4  text_vect_5  text_vect_6  text_vect_7  text_vect_8  \
0    -0.033911     0.004359      0.02092     -0.01589     0.022249   
1     0.007456     0.041769      0.03704      0.00225    -0.008879   

   text_vect_9  text_vect_10  text_vect_11  text_vect_12  text_vect_13  \
0     0.014778      0.015027     -0.035890     -0.002900      0.013295   
1    -0.009718      0.043835     -0.001604     -0.007495      0.049883   

   text_vect_14  text_vect_15  text_vect_16  text_vect_17  text_vect_18  \
0       0.04517      0.007890      0.000421      0.025048      0.014168   
1       0.02068      0.001809      0.029610     -0.053418      0.005492   

   text_vect_19  text_vect_20  text_vect_21  text_vect_22  text_vect_23  \
0     -0.042518     -0.039616     -0.005228     -0.006604      0.036101   
1      0.036592     -0.040964     -0.001432     -0.023874      0.004410   

   text_vect_24  text_vect_25  text_vect_26  text_vect_27  text_vect_28  \
0      0.042213      0.025370     -0.003305     -0.021069     -0.026389   
1      0.007027      0.020687      0.007595      0.010310     -0.017559   

   text_vect_29  text_vect_30  text_vect_31  text_vect_32  text_vect_33  \
0      0.034154     -0.017136      0.006048      0.020480     -0.019974   
1      0.033825      0.013476      0.039760      0.042392     -0.023853   

   text_vect_34  text_vect_35  text_vect_36  text_vect_37  text_vect_38  \
0     -0.001028     -0.010454      0.003524      0.037150      0.047859   
1     -0.016019      0.018561     -0.048938      0.002632      0.012729   

   text_vect_39  text_vect_40  text_vect_41  text_vect_42  text_vect_43  \
0     -0.068856     -0.017107     -0.001799     -0.001765     -0.019395   
1     -0.046141     -0.019764      0.005588     -0.022248     -0.003949   

   text_vect_44  text_vect_45  text_vect_46  text_vect_47  text_vect_48  \
0     -0.003731      0.016777     -0.008671      0.006507     -0.017094   
1     -0.078589     -0.007885     -0.000011     -0.007000      0.030419   

   text_vect_49  text_vect_50  text_vect_51  text_vect_52  text_vect_53  \
0      0.017093     -0.027028     -0.009333      0.038548      0.006556   
1      0.003075      0.033024      0.002380      0.003191     -0.042145   

   text_vect_54  text_vect_55  text_vect_56  text_vect_57  text_vect_58  \
0      0.027438     -0.007476     -0.024492      0.020877     -0.004892   
1      0.067408      0.025510      0.003584     -0.032802     -0.053132   

   text_vect_59  text_vect_60  text_vect_61  text_vect_62  text_vect_63  \
0     -0.006262      0.000765     -0.020052      0.003157      0.001788   
1      0.036182      0.001022      0.009138     -0.009701      0.002629   

   text_vect_64  text_vect_65  text_vect_66  text_vect_67  text_vect_68  \
0     -0.002731     -0.037521     -0.023036      0.020433      0.058812   
1      0.004591      0.003746     -0.048846     -0.005605      0.052039   

   text_vect_69  text_vect_70  text_vect_71  text_vect_72  text_vect_73  \
0     -0.022733     -0.007107     -0.029934     -0.007102      0.001971   
1     -0.014190      0.002171     -0.014709     -0.013629     -0.008634   

   text_vect_74  text_vect_75  text_vect_76  text_vect_77  text_vect_78  \
0     -0.013243     -0.028965      0.008166      0.023936     -0.

(6988, 779)

(6988, 50)

In [31]:
# создадим простую модель (MLP) на основе нейронной сети для обученяи на эмбеддингах LaBSE

# определяем MLP для многоклассовой задачи
class MLPModel(nn.Module):
    def __init__(self, input_dim, hidden_dims, output_dim):
        super(MLPModel, self).__init__()
        
        # задаем слои с разным количеством нейронов
        layers = []
        prev_dim = input_dim
        for hidden_dim in hidden_dims:
            layers.append(nn.Linear(prev_dim, hidden_dim))
            layers.append(nn.ReLU())
            prev_dim = hidden_dim
        self.hidden_layers = nn.Sequential(*layers)
        
        # выходной слой
        self.fc_out = nn.Linear(prev_dim, output_dim)
        
    def forward(self, x):
        x = self.hidden_layers(x)
        x = self.fc_out(x)
        return x

# классификатор с несколькими слоями
class MLPClassifier(BaseEstimator, ClassifierMixin):
    def __init__(self, input_dim, hidden_dims=[64, 128], output_dim=50, lr=0.001, epochs=10, batch_size=32):
        self.input_dim = input_dim
        self.hidden_dims = hidden_dims
        self.output_dim = output_dim
        self.lr = lr
        self.epochs = epochs
        self.batch_size = batch_size
        self.model = None
        self.classes_ = None  # Атрибут для хранения уникальных классов

    def _convert_to_numpy(self, X):
        """преобразуем входные данные в NumPy массив"""
        if isinstance(X, pd.DataFrame) or isinstance(X, pd.Series):
            return X.values
        elif isinstance(X, np.ndarray):
            return X
        else:
            raise ValueError("Неподдерживаемый формат данных. Необходимо передать DataFrame, Series, или массив NumPy.")

    def fit(self, X, y):
        X = self._convert_to_numpy(X)
        y = self._convert_to_numpy(y)

        X_tensor = torch.tensor(X, dtype=torch.float32)
        y_tensor = torch.tensor(y, dtype=torch.long)

        # создание модели
        self.model = MLPModel(self.input_dim, self.hidden_dims, self.output_dim)
        
        criterion = nn.CrossEntropyLoss()
        optimizer = optim.Adam(self.model.parameters(), lr=self.lr)

        for epoch in range(self.epochs):
            self.model.train()
            epoch_loss = 0
            for i in range(0, len(X_tensor), self.batch_size):
                batch_X = X_tensor[i:i + self.batch_size]
                batch_y = y_tensor[i:i + self.batch_size]

                optimizer.zero_grad()
                outputs = self.model(batch_X)
                loss = criterion(outputs, batch_y)
                loss.backward()
                optimizer.step()

                epoch_loss += loss.item()  # накопление потерь для эпохи

            # печатаем потерю после обработки всех батчей в эпохе
            #print(f"Epoch {epoch + 1}/{self.epochs}, Loss: {epoch_loss / (len(X_tensor) // self.batch_size):.4f}")

        # сохраняем уникальные классы после обучения модели
        self.classes_ = np.unique(y)

        return self

    def predict_proba(self, X):
        X = self._convert_to_numpy(X)
        self.model.eval()
        X_tensor = torch.tensor(X, dtype=torch.float32)
        with torch.no_grad():
            outputs = self.model(X_tensor)
            probabilities = torch.softmax(outputs, dim=1).numpy()
        return probabilities

    def predict(self, X):
        probabilities = self.predict_proba(X)
        return np.argmax(probabilities, axis=1)

    def score(self, X, y):
        X = self._convert_to_numpy(X)
        y = self._convert_to_numpy(y)
        predictions = self.predict(X)
        return (predictions == y).mean()

In [32]:
# создаём итоговый пайплайн с выбором модели и гиперпараметров для обучения на эмбеддингах LaBSE
pipe_final_bert = Pipeline([
    ('models', MultiOutputClassifier(LogisticRegression(random_state=RANDOM_STATE)))
])

# задаем наборы гиперпараметров для перебора
# аналогично моделям выше - оставил только лучшие наборы по итогам кросс-валидации, полный перебор опускаем

param_grid_bert = [
    # Словарь для модели LogisticRegression
    {
        'models': [MultiOutputClassifier(LogisticRegression(random_state=RANDOM_STATE))],
        'models__estimator__max_iter': [1000],
        'models__estimator__C': [10],
        'models__estimator__penalty': ['l2'],
        'models__estimator__class_weight': [None]
    },

    # Словарь для модели CatBoostClassifier
    {
        'models': [MultiOutputClassifier(CatBoostClassifier(silent=True, random_state=RANDOM_STATE, auto_class_weights='Balanced',
                                                            early_stopping_rounds=10))],
        'models__estimator__depth': [7],
        'models__estimator__iterations': [400],
        'models__estimator__learning_rate': [0.1],
        'models__estimator__l2_leaf_reg': [3],
        'models__estimator__border_count': [128],
        'models__estimator__rsm': [1],
        'models__estimator__subsample': [1],
        'models__estimator__random_strength': [1],
        'models__estimator__bagging_temperature': [0]
    },

    # Словарь для модели MLPClassifier
    {
        'models': [MultiOutputClassifier(MLPClassifier(input_dim=X_train_bert.shape[1], output_dim=50,))],
        'models__estimator__hidden_dims': [[256, 128, 64]],
        'models__estimator__lr': [1],
        'models__estimator__epochs': [10],
        'models__estimator__batch_size': [256]    
    }
    ]

In [33]:

# полный перебор гиперпараметров с помощью GridSearchCV с применением tqdm
if flag != 1:
    param_grid_bert = ParameterGrid(param_grid_bert)  # параметры для перебора внутри tqdm

    # настройка tqdm
    total_combinations = len(param_grid_bert)
    with tqdm(total=total_combinations) as pbar:
        best_score_bert = -float('inf')
        best_params_bert = None
        best_model_bert = None

        for params in param_grid_bert:
            grid_search_bert = GridSearchCV(
                pipe_final_bert,
                {k: [v] for k, v in params.items()},  # подставляем конкретный набор параметров
                cv=5,
                scoring='accuracy',
                n_jobs=-1,
            )
            model_bert = grid_search_bert.fit(X_train_bert, y_train)

            print(f'Метрика Accuracy: {model_bert.best_score_}')
            print(f'Параметры модели: {model_bert.best_params_}\n')

            # обновляем метрику
            if model_bert.best_score_ > best_score_bert:
                best_score_bert = model_bert.best_score_
                best_params_bert = params
                best_model_bert = model_bert

            pbar.update(1)  # обновляем прогресс бар tqdm

    # выгружаем лучшую метрику Accuracy и параметры модели
    print(f'Лучшая метрика Accuracy: {best_score_bert}')
    print(f'Параметры лучшей модели: {best_params_bert}')

  0%|          | 0/3 [00:00<?, ?it/s]

Метрика Accuracy: 0.5306280677069093
Параметры модели: {'models': MultiOutputClassifier(estimator=LogisticRegression(random_state=42)), 'models__estimator__C': 10, 'models__estimator__class_weight': None, 'models__estimator__max_iter': 1000, 'models__estimator__penalty': 'l2'}

Метрика Accuracy: 0.6252189701414128
Параметры модели: {'models': MultiOutputClassifier(estimator=<catboost.core.CatBoostClassifier object at 0x32e95b9d0>), 'models__estimator__bagging_temperature': 0, 'models__estimator__border_count': 128, 'models__estimator__depth': 7, 'models__estimator__iterations': 400, 'models__estimator__l2_leaf_reg': 3, 'models__estimator__learning_rate': 0.1, 'models__estimator__random_strength': 1, 'models__estimator__rsm': 1, 'models__estimator__subsample': 1}

Метрика Accuracy: 0.44390503664607284
Параметры модели: {'models': MultiOutputClassifier(estimator=MLPClassifier(input_dim=779)), 'models__estimator__batch_size': 256, 'models__estimator__epochs': 10, 'models__estimator__hidde

In [34]:
# сохранение лучшей модели
#joblib.dump(best_model_bert, 'best_model_clear_bert.pkl')
#print("Модель сохранена в файл best_model_clear_bert.pkl")

<a id='Проверка-лучшей-модели-на-валидационной-выборке'></a> 
## Проверка лучшей модели на валидационной выборке
[К содержанию](#К-содержанию)

Выберем лучшую модель, про итогам кросс-валиаци на разныех вариантах преобразовния текстов (TF-IDF, Word2Vec, LaBSE) и проведем проверку на валидационной выборке, ранее выделенной из тренировочной и не участвовавшей в обучении. Лучшие результаты с метрикой accuracy на уровне 0.62 показала модель, обученная Catboost на эмбеддингах предобученной LaBSE, ее и будем тестировать:

In [35]:
X_test_bert = X_test.copy()

tqdm.pandas()

# преобразуем текст в векторы
X_test_bert['text_vect'] = X_test_bert['text'].progress_apply(lambda x: model_LaBSE.encode(x))

# удаляем ненужные столбцы
X_test_bert = X_test_bert.drop(columns=['text'])

X_test_bert.head(2)

  0%|          | 0/1234 [00:00<?, ?it/s]

assessment  tags_assortment  tags_promotions  tags_delivery  tags_price  \
0         6.0                1                0              1           0   
1         5.0                1                1              1           1   

   tags_products_quality  tags_support  tags_catalog_navigation  tags_payment  \
0                      1             1                        1             0   
1                      1             1                        1             1   

   text_len  tags_count  \
0        29           5   
1         8           8   

                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  text_vect  
0  [0.00036013086, -0.03528022, -0.014954457, -0.016956663, -0.038329218, 0.015523258, 0.0032164063, 0.006735514, 0.03324024, 0.00019466922, 0.011180547, -0.0044298964, 0.0058418955, -0.014067804, 0.028526459, -0.034279082, -0.01368375, -0.006148947, -0.023980413, 0.0072173835, -0.0017660378, 0.00023879904, -0.026463639, 0.0040453374, 0.020404646, 0.019818928, 0.0060669715, -0.011822134, -0.020403316, 0.003350783, -0.002955692, -0.0148564745, 0.02083568, -0.018679205, 0.019180125, -0.015484745, -0.012740105, -0.007253425, 0.0028211174, -0.08572251, -0.01269066, -0.004179113, -0.032811362, -0.017562827, -0.06885304, -0.021773728, -0.025296947, 0.0021211137, 0.012182426, 0.017812537, -0.045365468, -0.033700168, 0.00066258555, -0.03489903, 0.047786444, -0.014547288, 0.0038958837, -0.026518002, -0.058763005, 0.044459797, -0.020373352, 0.0017497065, 0.005072954, -0.0031589556, 0.028559007, -0.03427765, -0.0063339598, 0.02292376, 0.04019004, -0.008439728, -0.028162451, -0.010550224, -0.0087918565, -0.002249871, -0.003329785, -0.0016154009, 0.015256508, 0.01902247, -0.035366297, 0.0022000219, -0.02113619, 0.01716768, 0.013021038, 0.06080408, 0.01723275, 0.017695704, 0.005156194, -0.00017913237, 0.006565541, 0.0038728684, 0.004810301, -0.006106544, -0.021511126, 0.0647173, -0.003998983, -0.006218472, -0.03855435, -0.012560402, 0.016796058, 0.009137364, ...]  
1                     [0.0014719601, 0.003666016, 0.03653107, 0.012741787, -0.029573008, 0.04964126, 0.045388594, -0.0022070406, 0.02124903, 0.020281028, 0.03251101, -0.0060531427, -0.029702775, 0.012295499, 0.023246251, -0.029331433, -0.021245703, -0.072890855, -0.020343602, 0.0223762, -0.0005439848, -0.032734904, -0.005439613, -0.017454714, 0.029549953, 0.03015772, -0.023922894, 0.014462702, -0.023643294, -0.00030633475, 0.011224688, -0.03319629, 0.04012765, -0.025069052, -0.023498459, -0.00527323, -0.012405067, -0.009468706, -0.009173998, -0.028914195, 0.0068997606, -0.009306354, 0.029397132, -0.02375635, -0.044040058, -0.029295627, -0.0221922, -0.020300273, -0.021854483, 0.04127881

In [36]:
# определяем количество столбцов в векторе
vector_length = len(X_test_bert['text_vect'].iloc[0])

# генерируем названия столбцов
column_names = [f"text_vect_{i}" for i in range(vector_length)]

# создаем DataFrame с заданными названиями столбцов
text_vect_df = pd.DataFrame(X_test_bert['text_vect'].tolist(), 
                            index=X_test_bert.index, 
                            columns=column_names)

# объединяем с исходным DataFrame
X_test_bert = pd.concat([X_test_bert, text_vect_df], axis=1)

# удаляем колонку с исходным вектором
X_test_bert = X_test_bert.drop(columns=['text_vect'])

In [37]:
display(X_test_bert.head(2))
display(X_test_bert.shape)
y_test.shape

assessment  tags_assortment  tags_promotions  tags_delivery  tags_price  \
0         6.0                1                0              1           0   
1         5.0                1                1              1           1   

   tags_products_quality  tags_support  tags_catalog_navigation  tags_payment  \
0                      1             1                        1             0   
1                      1             1                        1             1   

   text_len  tags_count  text_vect_0  text_vect_1  text_vect_2  text_vect_3  \
0        29           5     0.000360    -0.035280    -0.014954    -0.016957   
1         8           8     0.001472     0.003666     0.036531     0.012742   

   text_vect_4  text_vect_5  text_vect_6  text_vect_7  text_vect_8  \
0    -0.038329     0.015523     0.003216     0.006736     0.033240   
1    -0.029573     0.049641     0.045389    -0.002207     0.021249   

   text_vect_9  text_vect_10  text_vect_11  text_vect_12  text_vect_13  \
0     0.000195      0.011181     -0.004430      0.005842     -0.014068   
1     0.020281      0.032511     -0.006053     -0.029703      0.012295   

   text_vect_14  text_vect_15  text_vect_16  text_vect_17  text_vect_18  \
0      0.028526     -0.034279     -0.013684     -0.006149     -0.023980   
1      0.023246     -0.029331     -0.021246     -0.072891     -0.020344   

   text_vect_19  text_vect_20  text_vect_21  text_vect_22  text_vect_23  \
0      0.007217     -0.001766      0.000239     -0.026464      0.004045   
1      0.022376     -0.000544     -0.032735     -0.005440     -0.017455   

   text_vect_24  text_vect_25  text_vect_26  text_vect_27  text_vect_28  \
0      0.020405      0.019819      0.006067     -0.011822     -0.020403   
1      0.029550      0.030158     -0.023923      0.014463     -0.023643   

   text_vect_29  text_vect_30  text_vect_31  text_vect_32  text_vect_33  \
0      0.003351     -0.002956     -0.014856      0.020836     -0.018679   
1     -0.000306      0.011225     -0.033196      0.040128     -0.025069   

   text_vect_34  text_vect_35  text_vect_36  text_vect_37  text_vect_38  \
0      0.019180     -0.015485     -0.012740     -0.007253      0.002821   
1     -0.023498     -0.005273     -0.012405     -0.009469     -0.009174   

   text_vect_39  text_vect_40  text_vect_41  text_vect_42  text_vect_43  \
0     -0.085723     -0.012691     -0.004179     -0.032811     -0.017563   
1     -0.028914      0.006900     -0.009306      0.029397     -0.023756   

   text_vect_44  text_vect_45  text_vect_46  text_vect_47  text_vect_48  \
0     -0.068853     -0.021774     -0.025297      0.002121      0.012182   
1     -0.044040     -0.029296     -0.022192     -0.020300     -0.021854   

   text_vect_49  text_vect_50  text_vect_51  text_vect_52  text_vect_53  \
0      0.017813     -0.045365     -0.033700      0.000663     -0.034899   
1      0.041279      0.002115     -0.004531      0.010675     -0.017655   

   text_vect_54  text_vect_55  text_vect_56  text_vect_57  text_vect_58  \
0      0.047786     -0.014547      0.003896     -0.026518     -0.058763   
1      0.019110      0.021949      0.012645     -0.035178     -0.020414   

   text_vect_59  text_vect_60  text_vect_61  text_vect_62  text_vect_63  \
0      0.044460     -0.020373      0.001750      0.005073     -0.003159   
1      0.015851     -0.018897      0.030945      0.039899      0.012163   

   text_vect_64  text_vect_65  text_vect_66  text_vect_67  text_vect_68  \
0      0.028559     -0.034278     -0.006334      0.022924      0.040190   
1      0.044273     -0.032124     -0.038853     -0.008951      0.051184   

   text_vect_69  text_vect_70  text_vect_71  text_vect_72  text_vect_73  \
0     -0.008440     -0.028162     -0.010550     -0.008792     -0.002250   
1     -0.012568     -0.013888     -0.004758     -0.015794     -0.014115   

   text_vect_74  text_vect_75  text_vect_76  text_vect_77  text_vect_78  \
0     -0.003330     -0.001615      0.015257      0.019022     -0.

(1234, 779)

(1234, 50)

In [38]:
# делаем прогнозы на основе тестовых данных
X_test_pred = best_model_bert.predict(X_test_bert)

# считаем Accuracy для многометочной классификации
accuracy = accuracy_score(y_test, X_test_pred)

print('Метрика Accuracy на тестовых данных:', accuracy)

Метрика Accuracy на тестовых данных: 0.6523500810372771


<a id='Дообучение-модели-LaBSE-для-классификации'></a> 
## Дообучение модели LaBSE для классификации
[К содержанию](#К-содержанию)

Дополнительно попробуем проверить гипотезу с дообучением слоя классификации поверх самой модели LaBSE.

In [39]:
# скопируем тренировочный датасет для обработки под модель в случае необходимости
X_train_LaBSE_ft = X_train.copy()

In [40]:
# данная модель принимает предложения длиной, не больше 512 токенов, поэтому сначала проверю, какая максимальная длина в train и test:

seq_len_train = [len(str(i).split()) for i in X_train_LaBSE_ft['text']]
seq_len_test = [len(str(i).split()) for i in X_test['text']]
max_seq_len = max(max(seq_len_test), max(seq_len_train))
max_seq_len

52

In [41]:
# загружаем токенизатор и исходную модель LaBSE
tokenizer = AutoTokenizer.from_pretrained('sergeyzh/LaBSE-ru-turbo')
model_LaBSE_ft = AutoModel.from_pretrained('sergeyzh/LaBSE-ru-turbo')

In [42]:
# создаем модель с добавлением классификационного слоя

class LaBSEClassifier(nn.Module):
    def __init__(self, base_model, num_labels):
        super().__init__()
        self.base_model = base_model  # исходная модель
        self.output_layer = nn.Linear(768, num_labels)  # слой классификатора

    def forward(self, input_ids, attention_mask, labels=None):
        # получаем выходы из исходной модели
        outputs = self.base_model(input_ids=input_ids, attention_mask=attention_mask)
        pooled_output = outputs.last_hidden_state[:, 0]  # Используем выход [CLS] токена

        # пропускаем через классификационный слой
        logits = self.output_layer(pooled_output)

        # вычисляем функцию потерь
        loss = None
        if labels is not None:
            loss = F.binary_cross_entropy_with_logits(logits, labels)

        return {"loss": loss, "logits": logits} if loss is not None else logits
    
class TextDataset(torch.utils.data.Dataset):
    def __init__(self, encodings, labels):
        self.encodings = encodings
        self.labels = np.array(labels, dtype=np.float32)  # Используем float32 для работы с BCEWithLogitsLoss

    def __getitem__(self, idx):
        item = {
            'input_ids': self.encodings['input_ids'][idx].clone().detach(),
            'attention_mask': self.encodings['attention_mask'][idx].clone().detach(),
            'labels': torch.tensor(self.labels[idx], dtype=torch.float32)  # Метки в формате [batch_size, num_labels]
        }
        return item

    def __len__(self):
        return len(self.labels)

# функция для подготовки,токенизации данных
def prepare_data(texts, labels, tokenizer, test_size=0.2):
    
    assert len(texts) == len(labels), "Количество текстов и меток должно совпадать"
    
    texts = [str(text) for text in texts]
    labels = np.array(labels, dtype=np.float32)
    
    train_texts, val_texts, train_labels, val_labels = train_test_split(
        texts, labels, test_size=test_size, random_state=42, shuffle=True
    )
    
    train_encodings = tokenizer(
        train_texts,
        truncation=True,
        padding=True,
        max_length=512,
        return_tensors="pt"
    )
    
    val_encodings = tokenizer(
        val_texts,
        truncation=True,
        padding=True,
        max_length=512,
        return_tensors="pt"
    )
    
    train_dataset = TextDataset(train_encodings, train_labels)
    val_dataset = TextDataset(val_encodings, val_labels)
    
    return train_dataset, val_dataset

# функция расчета метрик
def compute_metrics(eval_pred):
    logits, labels = eval_pred
    
    # преобразуем логиты в вероятности
    probabilities = torch.sigmoid(torch.tensor(logits)).numpy()
    # а потом в бинарные метки с порогом 0.5
    predictions = (probabilities > 0.5).astype(int)
    
    # считаем метрики для многозадачного случая
    accuracy = (predictions == labels).mean()  # Доля правильных предсказаний (по всем задачам)
    
    return {"accuracy": accuracy}


In [43]:
# инициализируем модель
classifier_LaBSE_ft = LaBSEClassifier(model_LaBSE_ft, num_labels=50)

# готовим данные
train_LaBSE_ft, val_LaBSE_ft = prepare_data(X_train_LaBSE_ft['text'], y_train, tokenizer)

# задаем параметры дообучения
training_args = TrainingArguments(
    output_dir='./results',
    num_train_epochs=4,
    per_device_train_batch_size=8,
    per_device_eval_batch_size=8,
    eval_strategy='epoch',
    logging_strategy='epoch',
    save_strategy='epoch',
    logging_dir='./logs',
    load_best_model_at_end=True,
    metric_for_best_model='loss',
    greater_is_better=False,
    warmup_steps=100,
    weight_decay=0.01,
    learning_rate=1e-6
)

# инициализируем тренера
trainer = Trainer(
    model=classifier_LaBSE_ft,
    args=training_args,
    train_dataset=train_LaBSE_ft,
    eval_dataset=val_LaBSE_ft,
    compute_metrics=compute_metrics
)


In [44]:
# запускаем дообучение модели
if flag != 1:
    trainer.train()

    # выводим метрики
    print(trainer.evaluate())  

  0%|          | 0/2796 [00:00<?, ?it/s]

{'loss': 0.2843, 'grad_norm': 0.39627620577812195, 'learning_rate': 7.778189910979228e-07, 'epoch': 1.0}


  0%|          | 0/175 [00:00<?, ?it/s]

{'eval_loss': 0.12507055699825287, 'eval_accuracy': 0.9833905579399141, 'eval_runtime': 13.1778, 'eval_samples_per_second': 106.087, 'eval_steps_per_second': 13.28, 'epoch': 1.0}
{'loss': 0.1195, 'grad_norm': 0.3482030928134918, 'learning_rate': 5.185459940652819e-07, 'epoch': 2.0}


  0%|          | 0/175 [00:00<?, ?it/s]

{'eval_loss': 0.10808612406253815, 'eval_accuracy': 0.9833905579399141, 'eval_runtime': 13.9319, 'eval_samples_per_second': 100.345, 'eval_steps_per_second': 12.561, 'epoch': 2.0}
{'loss': 0.1079, 'grad_norm': 0.3000417649745941, 'learning_rate': 2.5927299703264093e-07, 'epoch': 3.0}


  0%|          | 0/175 [00:00<?, ?it/s]

{'eval_loss': 0.10168487578630447, 'eval_accuracy': 0.9833905579399141, 'eval_runtime': 14.0307, 'eval_samples_per_second': 99.638, 'eval_steps_per_second': 12.473, 'epoch': 3.0}
{'loss': 0.1035, 'grad_norm': 0.2780798673629761, 'learning_rate': 0.0, 'epoch': 4.0}


  0%|          | 0/175 [00:00<?, ?it/s]

{'eval_loss': 0.09973445534706116, 'eval_accuracy': 0.9833905579399141, 'eval_runtime': 15.0296, 'eval_samples_per_second': 93.016, 'eval_steps_per_second': 11.644, 'epoch': 4.0}
{'train_runtime': 1134.6937, 'train_samples_per_second': 19.706, 'train_steps_per_second': 2.464, 'train_loss': 0.1537844749308792, 'epoch': 4.0}


  0%|          | 0/175 [00:00<?, ?it/s]

{'eval_loss': 0.09973445534706116, 'eval_accuracy': 0.9833905579399141, 'eval_runtime': 15.9149, 'eval_samples_per_second': 87.842, 'eval_steps_per_second': 10.996, 'epoch': 4.0}


Поскольку кросс-валидация не проводилась при дообучении - проведем оценку на тестовой выборке, подготовим валидационные данные для тестирования дообученной модели.

In [45]:
# выбираем устройство (MPS, если доступен)
device = torch.device("mps") if torch.backends.mps.is_available() else torch.device("cpu")

# перемещаем модель на устройство
classifier_LaBSE_ft.to(device)

def predict_on_test(model, texts):
    model.eval()  # включаем режим оценки
    
    texts = [str(text) for text in texts]
    
    # делаем токенизацию входных текстов
    encodings = tokenizer(
        texts,
        truncation=True,
        padding=True,
        max_length=512,
        return_tensors="pt"
    )
    
    # перемещаем данных на устройство
    input_ids = encodings['input_ids'].to(device)
    attention_mask = encodings['attention_mask'].to(device)
    
    # получаем предсказания
    with torch.no_grad():
        logits = model(input_ids=input_ids, attention_mask=attention_mask)
    
    # преобразуем полученные логиты в метки с тем же порогом
    predictions = (torch.sigmoid(logits) > 0.5).long()
    return predictions.cpu().numpy()

In [46]:
# делаем прдсказания на валидационном наборе данных
X_test_pred = predict_on_test(classifier_LaBSE_ft, X_test['text'])

# считаем Accuracy для многометочной классификации
accuracy = accuracy_score(y_test, X_test_pred)

print('Метрика Accuracy на тестовых данных:', accuracy)

Метрика Accuracy на тестовых данных: 0.44408427876823336


<a id='Общий-вывод'></a> 
## Общий вывод
[К содержанию](#К-содержанию)


По итогу **загрузки данных** было определено, что тренировочный датасет данных невелик - 8708 строк, при этом наличествует порядка 250 пропусков и дубликатов, которые были сразу исключены. Также выявлено, что почти половина датасета не имеет признаков принадлежности ни к одному из целевых классов (все значения из 50 целевых классов нулевые) - исключать данные строки не стали для улучшения обучения эмбеддингов (т.к. некоторая часть срок действительно могла не принадлежать ни к одному из классов).  
Далее была проведена незначительная **предобработка данных**:  
- значения столбца tags были бинарно закодированы в 8 столбцов(8 уникальных значений);
- проведена лемматизация и очистка текстов комментариев от пунктуационных и служебных знаков;
- добавлены признаки длины комментария и количества тегов.
После этого данные были разделены на тренировочную и тестовую выборки с учетом стратификаци по "пустым" целевым значениям (см. выше).  

**Исследовательский анализ** данных не проводился, т.к. основным источником для классификации в искомых признаков являются тексты самих комментариев, единственный числовой признак - оценка доставки имеет сильно второстепенное значение для классификации комментария по одной из 50 категорий.  

Далее были сформированы несколько пайплайнов с нарастанием сложности обучаемых моделей:  

1. Сначала текст коментариев были векторизованы с помощью метода **TF-IDF**, на этих векторах, а также прочих исходных и добавленных признаках были обучены две модели с помощью базового алгоритма логистической регрессии (метрика accuracy на кросс-валидации составила 0.51) и более продвинутого алгоритма catboost (метрика accuracy - 0.52), обе модели показали минимальные значения на уровне baseline'а и совсем незначительно превысили случайный вариант распределения меток классов.
2. Далее на тренировочном датасете была обучена модель подготовки **эмбеддингов Word2Vec**, с помощью которой были сгенерированы эмбеддинги, на которых (а также прочих признаках) были обучены аналогичные модели с помощью логистической регрессии (метрика accuracy составила 0.48) и catboost (метрика accuracy - 0.56), эмбеддинги word2vec лучше представили связи в текстах, что помогло catboost качественне выявить закономерности для итоговой классификации.
3. Следующим этапом была использована **предобученная модель генерации эмбеддингов LaBSE** (на базе BERT), на эмбеддингах которой были обучены 3 типа моделей - с помощью логистической регрессии (метрика accuracy - 0.53), catboost (метрика accuracy - 0.62) и нейронная сеть MLP с тремя полносвязными слоями (метрика accuracy - 0.44). В данном случае модель обученная catboost значительно превзошла другие варианты.
4. Финальным этапом было проведено **дообучение исходной модели LaBSE** для классификации 50 целевых признаков (с добавлением поверх предобученной модели классификационного слоя), данная модель показала также низкий результат по метрике accuracy на уровне 0.44.  
   
Для проверки **на тестовом наборе** была отобрана **модель обученная catboost'ом на эмбеддингах LaBSE**, которая показала метрику **accuracy - 0.65**.

В заверешение анализа можно предположить, что улучшение метрики относительно baseline'a c 0.50 до 0.65 с учетом настолько незначительного датасета (6 тысяч строк, из которых половина с нулевыми целевыми классами, то есть фактически для обучения классификации по 50! классам использовались всего 3 тысячи строк) является неплохим результатом, а в целях улучшения результата можно рекомендовать:
- увеличить обучающую выборку;
- усовершенствовать нейронные сети для обучения моделей с помощью архитетур c памятью контекста (GRU/LSTM), а также специализированных архитектур для табличных данных (TabM);
- проанализировать и дополнить признаковое пространство - учитывая частоту ключевых слов в текстах, предварительной классификации "пустых" признаков с помощью предобученных моделей, также учета норы эмбеддингов, наличия/частоты специальных слов/эмодзи и пр.